x/shield/keeper/proposal.go

```diff
@@ -17,3 +21,339 @@ func (k Keeper) ClaimUnlock(ctx sdk.Context, proposalID, poolID uint64, loss sdk
+			remainingUbd := entry.Balance
+			if uncoveredPurchase.IsPositive() {
+				if uncoveredPurchase.GTE(entry.Balance) {
+					uncoveredPurchase = uncoveredPurchase.Sub(entry.Balance)
+					remainingUbd = sdk.ZeroInt()
+				} else {
+					remainingUbd = entry.Balance.Sub(uncoveredPurchase)
+					uncoveredPurchase = sdk.ZeroInt()
+				}
+			}
+
+			// Make payout after purchased is fully covered.
+			if remainingUbd.IsPositive() {
+				if remainingUbd.GTE(payout) {
+					k.PayFromUnbonding(ctx, ubd, payout)
+					return
+				}
+				k.PayFromUnbonding(ctx, ubd, remainingUbd)
+				payoutFromUnbonding = payoutFromUnbonding.Sub(remainingUbd)
+			}
+		}
+	}
+
+	// Update collateral and withdraw.
+	provider.Collateral = provider.Collateral.Sub(payout)
+	provider.Withdrawing = provider.Withdrawing.Sub(payout)
+	k.SetProvider(ctx, provider.Address, provider)
+
+	// Update withdraw queue.
+	withdraws := k.GetWithdrawQueueTimeSlice(ctx, withdraw.CompletionTime)
+	for i := range withdraws {
+		if withdraws[i].Address.Equals(withdraw.Address) && withdraws[i].Amount.Equal(withdraw.Amount) {
+			withdraws[i].Amount = withdraws[i].Amount.Sub(payout)
+			break
+		}
+	}
+	k.SetWithdrawQueueTimeSlice(ctx, withdraw.CompletionTime, withdraws)
+}
+
+// GetSortedUnbondingDelegations gets unbonding delegations sorted by completion time from latest to earliest.
+func (k Keeper) GetSortedUnbondingDelegations(ctx sdk.Context, delAddr sdk.AccAddress) []staking.UnbondingDelegation {
+	ubds := k.sk.GetAllUnbondingDelegations(ctx, delAddr)
+	var unbondingDelegations []staking.UnbondingDelegation
+	for _, ubd := range ubds {
+		for _, entry := range ubd.Entries {
+			unbondingDelegations = append(
+				unbondingDelegations,
+				types.NewUnbondingDelegation(ubd.DelegatorAddress, ubd.ValidatorAddress, entry),
+			)
+		}
+	}
+	sort.SliceStable(unbondingDelegations, func(i, j int) bool {
+		return unbondingDelegations[i].Entries[0].CompletionTime.After(unbondingDelegations[j].Entries[0].CompletionTime)
+	})
+	return unbondingDelegations
+}
+
+// MakePayoutByProvider undelegates delegations and send coins the the shield module.
+func (k Keeper) MakePayoutByProvider(ctx sdk.Context, providerAddr sdk.AccAddress, purchased, payout sdk.Int) error {
+	provider, found := k.GetProvider(ctx, providerAddr)
+	if !found {
+		return types.ErrProviderNotFound
+	}
+
+	// If collateral - withdraw >= purchased + payout, make payouts from collateral - withdraw.
+	// If purchased <= collateral - withdraw < purchased + payout, make payouts from collateral - withdraw and withdraw.
+	// Otherwise, make payouts from withdraw.
+	uncoveredPurchase := sdk.ZeroInt()
+	payoutFromCollateral := sdk.ZeroInt()
+	if provider.Collateral.Sub(provider.Withdrawing).GTE(purchased.Add(payout)) {
+		payoutFromCollateral = payout
+	} else if provider.Collateral.Sub(provider.Withdrawing).GTE(purchased) {
+		payoutFromCollateral = provider.Collateral.Sub(provider.Withdrawing).Sub(purchased)
```

Comment:
Shouldn't we pay as much as we can with collateral here instead of just subtracting `purchased`?